### PR TITLE
Update to Istio 1.0.0

### PR DESCRIPTION
**Pull request by rlenglet**
Upgrade the Envoy and Istio SHAs to the latest commit in https://github.com/istio/proxy/tree/1.0.0.
Update the generated Protobuf files.
Update the Cilium filters to Envoy's internal API changes.
Update the Vagrant box and cilium-builder Docker image to update the Bazel cache.  Install `ninja-build`, which is a new Envoy build dependency.
Update the proxy image dockerfiles to use Istio's new `proxytproxy` image.

Disable the ADS protocol support in the Cilium Agent, since it was unused.
Continuing to support ADS will require supporting Incremental xDS, cf. https://github.com/cilium/cilium/issues/5051.

Adapt Envoy route generation to Envoy's updated `HeaderMatcher` API.

Update the Istio GSG to Istio 1.0.0.
Enable the Pilot cilium plugin on the Pilot command line using the new `--plugins` flag.
Use the new Helm value to configure the Pilot image name.
Add `--validate=false` to every `kubectl` command to create deployments injected with Istio sidecar proxies, since kube-inject generates specs that are not 100% compatible with k8s 1.9.

Exit an endpoint's synchronous creation if it is in sidecar proxy mode.
If the endpoint is determined to have a sidecar proxy, return from the endpoint creation to let the sidecar container start, in case it is required to enforce L7 rules. Otherwise, endpoint creation would continuously timeout and the pod wouldn't start.

Signed-off-by: Romain Lenglet <romain@covalent.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4923)

### CODE REVIEW

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Unit tests failed:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/5158/execution/node/55/log/
```
01:34:02 pkg/envoy/grpc.go:53:69: cannot use dsServer (type *xdsGRPCServer) as type "github.com/cilium/cilium/pkg/envoy/envoy/service/discovery/v2".AggregatedDiscoveryServiceServer in argument to "github.com/cilium/cilium/pkg/envoy/envoy/service/discovery/v2".RegisterAggregatedDiscoveryServiceServer:
01:34:02 	*xdsGRPCServer does not implement "github.com/cilium/cilium/pkg/envoy/envoy/service/discovery/v2".AggregatedDiscoveryServiceServer (missing IncrementalAggregatedResources method)
01:34:02 pkg/envoy/server.go:279:83: unknown field 'Value' in struct literal of type route.HeaderMatcher
01:34:02 pkg/envoy/server.go:279:98: unknown field 'Regex' in struct literal of type route.HeaderMatcher
01:34:02 pkg/envoy/server.go:283:85: unknown field 'Value' in struct literal of type route.HeaderMatcher
01:34:02 pkg/envoy/server.go:283:102: unknown field 'Regex' in struct literal of type route.HeaderMatcher
01:34:02 pkg/envoy/server.go:291:88: unknown field 'Value' in struct literal of type route.HeaderMatcher
01:34:02 pkg/envoy/server.go:291:103: unknown field 'Regex' in struct literal of type route.HeaderMatcher
01:34:02 pkg/envoy/server.go:307:80: unknown field 'Value' in struct literal of type route.HeaderMatcher
01:34:02 pkg/envoy/sort.go:218:9: m1.Value undefined (type *route.HeaderMatcher has no field or method Value)
01:34:02 pkg/envoy/sort.go:218:20: m2.Value undefined (type *route.HeaderMatcher has no field or method Value)
01:34:02 pkg/envoy/sort.go:218:20: too many errors
01:34:03 FAIL	github.com/cilium/cilium/daemon [build failed]
```
This is due to a change in the Envoy API.

File: Vagrantfile
Comment (eloycoto) [inline Vagrantfile]: Quick one, if you can use 103 here will be great.

File: test/Vagrantfile
Comment (eloycoto) [inline test/Vagrantfile]: Quick one, if you can use 103 here will be great.

**Comment by jrajahalme** [comment]
All changes see logically sound, and envoy builds. make in the main repo fails, however:
```
go tool vet api pkg envoy plugins bpf cilium daemon monitor cilium-health bugtool
pkg/endpoint/endpoint.go:2555: arg entry.Key for printf verb %s of wrong type: github.com/cilium/cilium/pkg/maps/policymap.PolicyKey
pkg/endpoint/endpoint.go:2568: arg keyToAdd for printf verb %s of wrong type: github.com/cilium/cilium/pkg/maps/policymap.PolicyKey
pkg/envoy/server.go:278: github.com/cilium/cilium/pkg/envoy/envoy/api/v2/route.HeaderMatcher_RegexMatch composite literal uses unkeyed fields
pkg/envoy/server.go:283: github.com/cilium/cilium/pkg/envoy/envoy/api/v2/route.HeaderMatcher_RegexMatch composite literal uses unkeyed fields
pkg/envoy/server.go:292: github.com/cilium/cilium/pkg/envoy/envoy/api/v2/route.HeaderMatcher_RegexMatch composite literal uses unkeyed fields
pkg/envoy/server.go:309: github.com/cilium/cilium/pkg/envoy/envoy/api/v2/route.HeaderMatcher_ExactMatch composite literal uses unkeyed fields
pkg/envoy/server.go:314: github.com/cilium/cilium/pkg/envoy/envoy/api/v2/route.HeaderMatcher_PresentMatch composite literal uses unkeyed fields
cilium/cmd/endpoint_config.go:87: arg value for printf verb %d of wrong type: bool
daemon/main.go:680: possible formatting directive in Fatal call
daemon/main.go:688: possible formatting directive in Fatal call
Makefile:212: recipe for target 'govet' failed
make: *** [govet] Error 1
```

**Comment by rlenglet** [comment]
I need to regenerate the Vagrant box, due to changes in the docker-ce package that is provisionned. Blocked on https://github.com/cilium/packer-ci-build/pull/77.

**Comment by rlenglet** [comment]
@jrajahalme Weird, that `go vet` command doesn't fail for me in the Vagrant VM on the HEAD of this PR.

**Comment by jrajahalme** [comment]
`make -C envoy tests` fails, or hangs, and then times out. Something in the integration testing infra has changed, and the tests keep on waiting for a response, which never comes. Seen this happen before, I'll look into this.

**Comment by rlenglet** [comment]
@jrajahalme I think you're running a different version of go vet. I've seen that too outside of the Vagrant VM. Especially this issue, which is definitely bogus:
```
cilium/cmd/endpoint_config.go:87: arg value for printf verb %d of wrong type: bool
```
I've fixed the `composite literal uses unkeyed fields` issues.

**Comment by jrajahalme** [comment]
integration test needs header matcher adjustments:
```
diff --git a/envoy/cilium_integration_test.cc b/envoy/cilium_integration_test.cc
index 946201d0c..fe8d74fd4 100644
--- a/envoy/cilium_integration_test.cc
+++ b/envoy/cilium_integration_test.cc
@@ -48,30 +48,30 @@ resources:
     - remote_policies: [ 1 ]
       http_rules:
         http_rules:
-        - headers: [ { name: ':path', value: '/allowed', regex: false } ]
-        - headers: [ { name: ':path', value: '.*public$', regex: true } ]
-        - headers: [ { name: ':authority', value: 'allowedHOST', regex: false } ]
-        - headers: [ { name: ':authority', value: '.*REGEX.*', regex: true } ]
-        - headers: [ { name: ':method', value: 'PUT', regex: false }, { name: ':path', value: '/public/opinions', regex: false } ]
+        - headers: [ { name: ':path', exact_match: '/allowed' } ]
+        - headers: [ { name: ':path', regex_match: '.*public$' } ]
+        - headers: [ { name: ':authority', exact_match: 'allowedHOST' } ]
+        - headers: [ { name: ':authority', regex_match: '.*REGEX.*' } ]
+        - headers: [ { name: ':method', exact_match: 'PUT' }, { name: ':path', exact_match: '/public/opinions' } ]
     - remote_policies: [ 2 ]
       http_rules:
         http_rules:
-        - headers: [ { name: ':path', value: '/only-2-allowed', regex: false } ]
+        - headers: [ { name: ':path', exact_match: '/only-2-allowed' } ]
   egress_per_port_policies:
   - port: 80
     rules:
     - remote_policies: [ 1 ]
       http_rules:
         http_rules:
-        - headers: [ { name: ':path', value: '/allowed', regex: false } ]
-        - headers: [ { name: ':path', value: '.*public$', regex: true } ]
-        - headers: [ { name: ':authority', value: 'allowedHOST', regex: false } ]
-        - headers: [ { name: ':authority', value: '.*REGEX.*', regex: true } ]
-        - headers: [ { name: ':method', value: 'PUT', regex: false }, { name: ':path', value: '/public/opinions', regex: false } ]
+        - headers: [ { name: ':path', exact_match: '/allowed' } ]
+        - headers: [ { name: ':path', regex_match: '.*public$' } ]
+        - headers: [ { name: ':authority', exact_match: 'allowedHOST' } ]
+        - headers: [ { name: ':authority', regex_match: '.*REGEX.*' } ]
+        - headers: [ { name: ':method', exact_match: 'PUT' }, { name: ':path', exact_match: '/public/opinions' } ]
     - remote_policies: [ 2 ]
       http_rules:
         http_rules:
-        - headers: [ { name: ':path', value: '/only-2-allowed', regex: false } ]
+        - headers: [ { name: ':path', exact_match: '/only-2-allowed' } ]
 )EOF";
   
 namespace Filter {
@@ -648,7 +648,7 @@ TEST_P(CiliumIntegrationTest, DuplicatePort) {
     - remote_policies: [ 2 ]
       http_rules:
         http_rules:
-        - headers: [ { name: ':path', value: '/only-2-allowed', regex: false } ]
+        - headers: [ { name: ':path', exact_match: '/only-2-allowed' } ]
 )EOF";
 
   // This would normally be allowed, but since the policy fails, everything will be rejected.
```

**Comment by jrajahalme** [comment]
@rlenglet Funny I get the `go vet` errors *within* the just bootstrapped vagrantbox!

`go version` returns:
```
vagrant@runtime1:~/go/src/github.com/cilium/cilium$ go version
go version go1.9.3 linux/amd64
vagrant@runtime1:~/go/src/github.com/cilium/cilium$ which go 
/usr/local/bin/go
```
I though I saw upgrades to newer go versions some time ago?

**Comment by jrajahalme** [comment]
OK, I had not destroyed the VM image, trying again...

**Comment by rlenglet** [comment]
> integration test needs header matcher adjustments:

Integrated those changes. Thank you!

**Comment by rlenglet** [comment]
BTW, `tests/10-proxy.sh` passes.

**Comment by rlenglet** [comment]
Updated the Vagrant box to [version 105](https://jenkins.cilium.io/job/Vagrant-Master-Boxes-Packer-Build/105/).

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Updated Envoy and Istio Proxy filters to the `1.0.0` release tag.
There's a new Envoy build dependency (`ninja-build`), which will require updating the Vagrant VM again, cf. https://github.com/cilium/packer-ci-build/pull/79.

**Comment by rlenglet** [comment]
test-docs-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-docs-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-docs-please

**Comment by rlenglet** [comment]
Verified manually that the Istio GSG now works well with Istio 1.0.0.
Please review.

**Comment by rlenglet** [comment]
I will also try enabling mTLS in the Istio GSG, but will probably submit that as a separate PR if it works.

**Comment by rlenglet** [comment]
Docs build passed: https://jenkins.cilium.io/job/Cilium-PR-Doc-Tests/410/

**Review by ianvernon — APPROVED** [review]

**Comment by rlenglet** [comment]
@jrajahalme after switching to the Envoy and Istio filter SHAs used in Istio 1.0.0, `cilium_integration_test` is passing. It's all good.
```
Target //:cilium_integration_test up-to-date:
  bazel-bin/cilium_integration_test
INFO: Elapsed time: 214.110s, Critical Path: 22.62s
INFO: 54 processes: 53 linux-sandbox, 1 local.
INFO: Build completed successfully, 58 total actions
//:cilium_integration_test                                               PASSED in 3.6s

Executed 1 out of 1 test: 1 test passes.
```

**Review by eloycoto — APPROVED** [review]

**Comment by rlenglet** [comment]
Successful CI build: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/5408/

**Review by tgraf — APPROVED** [review]

File: daemon/endpoint.go
Comment (raybejjani) [inline daemon/endpoint.go]: Could you add a log here explaining the early return? I would advocate for an Info, since it's violating the synchronous wait flag.

Is there no other place to make this check? the `e.hasSidecarProxy` flag seems to be based only on the labels from k8s. Is it already available when `cilium-cni` is invoked? If so, could we then not use the synchronous option in that case?

File: pkg/envoy/sort.go
Comment (raybejjani) [inline pkg/envoy/sort.go:214]: Since this function is now quite a bit more complex, could you add a comment explaining the general logic? The getters are autogenerated and I'm not sure how to figure out what they fields are actually for.

**Review by raybejjani — DISMISSED** [review]
Mostly fine. Neither of my comments is particularly blocking, in case you want to merge this before I get a chance to approve.

File: daemon/endpoint.go
Comment (rlenglet) [reply daemon/endpoint.go]: Added an info log statement.

Yes, the sidecar proxy mode is determined based on k8s labels. Since the synchronous request will wait for up to 55 seconds, it gives enough time for the daemon to update the endpoint's labels. It typically happens within a few seconds after the endpoint creation, in a separate controller. So before the synchronous request times out, we'll know whether the endpoint is in sidecar mode.

File: pkg/envoy/sort.go
Comment (rlenglet) [reply pkg/envoy/sort.go:214]: Added this comment:
```
	// Compare the header_match_specifier oneof field, by comparing each
	// possible field in the oneof individually:
	// - exactMatch
	// - regexMatch
	// - rangeMatch
	// - presentMatch
	// - prefixMatch
	// - suffixMatch
	// Use the getters to access the fields and return zero values when they
	// are not set.
```

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-docs-please

**Comment by jrajahalme** [comment]
Test/Build failure:
```
16:07:44 docker-compose -f test/docker-compose.yml -p $JOB_BASE_NAME-$BUILD_NUMBER run --rm precheck
16:07:44 /usr/local/bin/docker-compose: 2: /usr/local/bin/docker-compose: Syntax error: newline unexpected
16:07:44 Makefile:29: recipe for target 'jenkins-precheck' failed
16:07:44 make: *** [jenkins-precheck] Error 2
```

**Review by jrajahalme — APPROVED** [review]

**Comment by ianvernon** [comment]
The build failure occurred on master as well (https://jenkins.cilium.io/job/Ginkgo-CI-Tests-Pipeline/3508/ ). Appears to be infrastructure related. The build for this PR which failed also ran on the same node (jenkins-node-1).

**Comment by ianvernon** [comment]
The build failure occurred on master as well (https://jenkins.cilium.io/job/Ginkgo-CI-Tests-Pipeline/3508/ ). Appears to be infrastructure related. The build for this PR which failed also ran on the same node (jenkins-node-1). This node has been deleted by our Jenkins infrastructure, so we don't have to clean it up ourselves.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/envoy/sort.go
Comment (raybejjani) [reply pkg/envoy/sort.go:214]: Excellent, exactly what I needed :) Thanks

File: daemon/endpoint.go
Comment (raybejjani) [reply daemon/endpoint.go]: Oh, I see. I thought these labels were set on endpoint create. If we wait on the labels from k8s I see how this would wait until they arrive, which will be some time after the endpoint create. I guess in that sense it isn't really violating the synchronous flag since it still waits.

**Pull request merged**
